File: sdk/typescript/src/namespaces/ideas.ts
/**
 * Ideas Namespace API
 *
 * Provides endpoints for the Idea Canvas (Stage 1 of the
 * Idea-to-Execution Pipeline): canvas CRUD, node/edge management,
 * React Flow export, and promotion to goals.
 */

import type { AragoraClient } from '../client';

/** An idea node on the canvas */
export interface IdeaNode {
  id: string;
  label: string;
  idea_type: string;
  position?: { x: number; y: number };
  data?: Record<string, unknown>;
}

/** An edge connecting two idea nodes */
export interface IdeaEdge {
  id: string;
  source_id: string;
  target_id: string;
  edge_type?: string;
  label?: string;
  data?: Record<string, unknown>;
}

/** An idea canvas document */
export interface IdeaCanvas {
  id: string;
  name: string;
  description?: string;
  owner_id?: string;
  workspace_id?: string;
  metadata?: Record<string, unknown>;
  nodes?: IdeaNode[];
  edges?: IdeaEdge[];
  created_at: string;
  updated_at: string;
}

/** Result of promoting nodes to goals */
export interface PromotionResult {
  goals_canvas: Record<string, unknown>;
  provenance: Record<string, unknown>[];
  promoted_count: number;
}

/**
 * Ideas namespace for the Idea Canvas (Stage 1 of the
 * Idea-to-Execution Pipeline).
 *
 * @example
 * ```typescript
 * // Canvas CRUD
 * const canvases = await client.ideas.listCanvases();
 * const canvas = await client.ideas.createCanvas('Brainstorm');
 *
 * // Add nodes and edges
 * const node = await client.ideas.addNode(canvas.id, 'Rate Limiter');
 * const edge = await client.ideas.addEdge(canvas.id, nodeA.id, nodeB.id);
 *
 * // Export & promote
 * const exported = await client.ideas.exportCanvas(canvas.id);
 * const result = await client.ideas.promoteNodes(canvas.id, [node.id]);
 * ```
 */
export class IdeasNamespace {
  constructor(private client: AragoraClient) {}

  // -------------------------------------------------------------------
  // Canvas CRUD
  // -------------------------------------------------------------------

  /** List idea canvases with optional filtering. */
  async listCanvases(options?: {
    workspace_id?: string;
    owner_id?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ canvases: IdeaCanvas[] }> {
    return this.client.request<{ canvases: IdeaCanvas[] }>(
      'GET',
      '/api/v1/ideas',
      { params: options },
    );
  }

  /** Create a new idea canvas. */
  async createCanvas(
    name: string,
    description: string = '',
    metadata?: Record<string, unknown>,
  ): Promise<IdeaCanvas> {
    const body: Record<string, unknown> = { name, description };
    if (metadata !== undefined) {
      body.metadata = metadata;
    }
    return this.client.request<IdeaCanvas>('POST', '/api/v1/ideas', { body });
  }

  /** Get a canvas by ID, including its nodes and edges. */
  async getCanvas(canvasId: string): Promise<IdeaCanvas> {
    return this.client.request<IdeaCanvas>(
      'GET',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}`,
    );
  }

  /** Update canvas metadata. */
  async updateCanvas(
    canvasId: string,
    updates: {
      name?: string;
      description?: string;
      metadata?: Record<string, unknown>;
    },
  ): Promise<IdeaCanvas> {
    return this.client.request<IdeaCanvas>(
      'PUT',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}`,
      { body: updates },
    );
  }

  /** Delete a canvas and all its nodes and edges. */
  async deleteCanvas(canvasId: string): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Node CRUD
  // -------------------------------------------------------------------

  /** Add a node to a canvas. */
  async addNode(
    canvasId: string,
    label: string,
    ideaType: string = 'concept',
    position?: { x: number; y: number },
    data?: Record<string, unknown>,
  ): Promise<IdeaNode> {
    const body: Record<string, unknown> = { label, idea_type: ideaType };
    if (position !== undefined) {
      body.position = position;
    }
    if (data !== undefined) {
      body.data = data;
    }
    return this.client.request<IdeaNode>(
      'POST',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/nodes`,
      { body },
    );
  }

  /** Update a node on a canvas. */
  async updateNode(
    canvasId: string,
    nodeId: string,
    updates: {
      label?: string;
      position?: { x: number; y: number };
      data?: Record<string, unknown>;
    },
  ): Promise<IdeaNode> {
    return this.client.request<IdeaNode>(
      'PUT',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/nodes/${encodeURIComponent(nodeId)}`,
      { body: updates },
    );
  }

  /** Delete a node from a canvas. */
  async deleteNode(
    canvasId: string,
    nodeId: string,
  ): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/nodes/${encodeURIComponent(nodeId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Edge CRUD
  // -------------------------------------------------------------------

  /** Add an edge connecting two nodes on a canvas. */
  async addEdge(
    canvasId: string,
    sourceId: string,
    targetId: string,
    edgeType: string = 'default',
    label: string = '',
    data?: Record<string, unknown>,
  ): Promise<IdeaEdge> {
    const body: Record<string, unknown> = {
      source_id: sourceId,
      target_id: targetId,
      edge_type: edgeType,
      label,
    };
    if (data !== undefined) {
      body.data = data;
    }
    return this.client.request<IdeaEdge>(
      'POST',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/edges`,
      { body },
    );
  }

  /** Delete an edge from a canvas. */
  async deleteEdge(
    canvasId: string,
    edgeId: string,
  ): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/edges/${encodeURIComponent(edgeId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Export & Promote
  // -------------------------------------------------------------------

  /** Export a canvas as React Flow JSON (nodes + edges arrays). */
  async exportCanvas(canvasId: string): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'GET',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/export`,
    );
  }

  /** Promote selected nodes from the idea canvas to the goals stage. */
  async promoteNodes(
    canvasId: string,
    nodeIds: string[],
  ): Promise<PromotionResult> {
    return this.client.request<PromotionResult>(
      'POST',
      `/api/v1/ideas/${encodeURIComponent(canvasId)}/promote`,
      { body: { node_ids: nodeIds } },
    );
  }
}

File: sdk/python/aragora_sdk/namespaces/ideas.py
"""
Ideas Namespace API

Provides methods for the Idea Canvas (Stage 1 of the Idea-to-Execution Pipeline):
- Canvas CRUD (list, create, get, update, delete)
- Node CRUD (add, update, delete)
- Edge CRUD (add, delete)
- Export to React Flow JSON
- Promote nodes to goals
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class IdeasAPI:
    """Synchronous Ideas Canvas API."""

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # -------------------------------------------------------------------
    # Canvas CRUD
    # -------------------------------------------------------------------

    def list_canvases(
        self,
        workspace_id: str | None = None,
        owner_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List idea canvases with optional filtering.

        Args:
            workspace_id: Filter by workspace.
            owner_id: Filter by owner.
            limit: Maximum number of canvases to return.
            offset: Number of canvases to skip for pagination.

        Returns:
            Dictionary with ``canvases`` list and pagination metadata.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        if owner_id is not None:
            params["owner_id"] = owner_id
        return self._client._request("GET", "/api/v1/ideas", params=params)

    def create_canvas(
        self,
        name: str,
        description: str = "",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a new idea canvas.

        Args:
            name: Canvas display name.
            description: Optional description of the canvas purpose.
            metadata: Arbitrary key-value metadata.

        Returns:
            The newly created canvas object.
        """
        body: dict[str, Any] = {"name": name, "description": description}
        if metadata is not None:
            body["metadata"] = metadata
        return self._client._request("POST", "/api/v1/ideas", json=body)

    def get_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Get a canvas by ID, including its nodes and edges.

        Args:
            canvas_id: Unique canvas identifier.

        Returns:
            Full canvas object with nodes, edges, and metadata.
        """
        return self._client._request("GET", f"/api/v1/ideas/{canvas_id}")

    def update_canvas(
        self,
        canvas_id: str,
        name: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update canvas metadata.

        Args:
            canvas_id: Canvas to update.
            name: New display name (if provided).
            description: New description (if provided).
            metadata: New metadata (if provided).

        Returns:
            Updated canvas object.
        """
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if metadata is not None:
            body["metadata"] = metadata
        return self._client._request("PUT", f"/api/v1/ideas/{canvas_id}", json=body)

    def delete_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Delete a canvas and all its nodes and edges.

        Args:
            canvas_id: Canvas to delete.

        Returns:
            Deletion confirmation with ``success`` flag.
        """
        return self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}")

    # -------------------------------------------------------------------
    # Node CRUD
    # -------------------------------------------------------------------

    def add_node(
        self,
        canvas_id: str,
        label: str,
        idea_type: str = "concept",
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add a node to a canvas.

        Args:
            canvas_id: Target canvas.
            label: Node display label.
            idea_type: Node type (e.g. ``concept``, ``question``, ``evidence``).
            position: Optional ``{x, y}`` position on the canvas.
            data: Arbitrary node data payload.

        Returns:
            The newly created node object.
        """
        body: dict[str, Any] = {"label": label, "idea_type": idea_type}
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return self._client._request("POST", f"/api/v1/ideas/{canvas_id}/nodes", json=body)

    def update_node(
        self,
        canvas_id: str,
        node_id: str,
        label: str | None = None,
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update a node on a canvas.

        Args:
            canvas_id: Canvas containing the node.
            node_id: Node to update.
            label: New label (if provided).
            position: New position (if provided).
            data: New data payload (if provided).

        Returns:
            Updated node object.
        """
        body: dict[str, Any] = {}
        if label is not None:
            body["label"] = label
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return self._client._request("PUT", f"/api/v1/ideas/{canvas_id}/nodes/{node_id}", json=body)

    def delete_node(self, canvas_id: str, node_id: str) -> dict[str, Any]:
        """Delete a node from a canvas.

        Args:
            canvas_id: Canvas containing the node.
            node_id: Node to delete.

        Returns:
            Deletion confirmation.
        """
        return self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}/nodes/{node_id}")

    # -------------------------------------------------------------------
    # Edge CRUD
    # -------------------------------------------------------------------

    def add_edge(
        self,
        canvas_id: str,
        source_id: str,
        target_id: str,
        edge_type: str = "default",
        label: str = "",
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add an edge connecting two nodes on a canvas.

        Args:
            canvas_id: Target canvas.
            source_id: Source node ID.
            target_id: Target node ID.
            edge_type: Edge classification (default ``"default"``).
            label: Optional edge label.
            data: Arbitrary edge data payload.

        Returns:
            The newly created edge object.
        """
        body: dict[str, Any] = {
            "source_id": source_id,
            "target_id": target_id,
            "edge_type": edge_type,
            "label": label,
        }
        if data is not None:
            body["data"] = data
        return self._client._request("POST", f"/api/v1/ideas/{canvas_id}/edges", json=body)

    def delete_edge(self, canvas_id: str, edge_id: str) -> dict[str, Any]:
        """Delete an edge from a canvas.

        Args:
            canvas_id: Canvas containing the edge.
            edge_id: Edge to delete.

        Returns:
            Deletion confirmation.
        """
        return self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}/edges/{edge_id}")

    # -------------------------------------------------------------------
    # Export & Promote
    # -------------------------------------------------------------------

    def export_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Export a canvas as React Flow JSON.

        Args:
            canvas_id: Canvas to export.

        Returns:
            React Flow format with ``nodes`` and ``edges`` arrays.
        """
        return self._client._request("GET", f"/api/v1/ideas/{canvas_id}/export")

    def promote_nodes(self, canvas_id: str, node_ids: list[str]) -> dict[str, Any]:
        """Promote selected nodes from the idea canvas to the goals stage.

        Args:
            canvas_id: Canvas containing the nodes.
            node_ids: List of node IDs to promote.

        Returns:
            Promotion result with ``goals_canvas``, ``provenance``,
            and ``promoted_count``.
        """
        return self._client._request(
            "POST",
            f"/api/v1/ideas/{canvas_id}/promote",
            json={"node_ids": node_ids},
        )


class AsyncIdeasAPI:
    """Asynchronous Ideas Canvas API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # -------------------------------------------------------------------
    # Canvas CRUD
    # -------------------------------------------------------------------

    async def list_canvases(
        self,
        workspace_id: str | None = None,
        owner_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List idea canvases with optional filtering."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        if owner_id is not None:
            params["owner_id"] = owner_id
        return await self._client._request("GET", "/api/v1/ideas", params=params)

    async def create_canvas(
        self,
        name: str,
        description: str = "",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a new idea canvas."""
        body: dict[str, Any] = {"name": name, "description": description}
        if metadata is not None:
            body["metadata"] = metadata
        return await self._client._request("POST", "/api/v1/ideas", json=body)

    async def get_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Get a canvas by ID, including its nodes and edges."""
        return await self._client._request("GET", f"/api/v1/ideas/{canvas_id}")

    async def update_canvas(
        self,
        canvas_id: str,
        name: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update canvas metadata."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if metadata is not None:
            body["metadata"] = metadata
        return await self._client._request("PUT", f"/api/v1/ideas/{canvas_id}", json=body)

    async def delete_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Delete a canvas and all its nodes and edges."""
        return await self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}")

    # -------------------------------------------------------------------
    # Node CRUD
    # -------------------------------------------------------------------

    async def add_node(
        self,
        canvas_id: str,
        label: str,
        idea_type: str = "concept",
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add a node to a canvas."""
        body: dict[str, Any] = {"label": label, "idea_type": idea_type}
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return await self._client._request("POST", f"/api/v1/ideas/{canvas_id}/nodes", json=body)

    async def update_node(
        self,
        canvas_id: str,
        node_id: str,
        label: str | None = None,
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update a node on a canvas."""
        body: dict[str, Any] = {}
        if label is not None:
            body["label"] = label
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return await self._client._request(
            "PUT", f"/api/v1/ideas/{canvas_id}/nodes/{node_id}", json=body
        )

    async def delete_node(self, canvas_id: str, node_id: str) -> dict[str, Any]:
        """Delete a node from a canvas."""
        return await self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}/nodes/{node_id}")

    # -------------------------------------------------------------------
    # Edge CRUD
    # -------------------------------------------------------------------

    async def add_edge(
        self,
        canvas_id: str,
        source_id: str,
        target_id: str,
        edge_type: str = "default",
        label: str = "",
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add an edge connecting two nodes on a canvas."""
        body: dict[str, Any] = {
            "source_id": source_id,
            "target_id": target_id,
            "edge_type": edge_type,
            "label": label,
        }
        if data is not None:
            body["data"] = data
        return await self._client._request("POST", f"/api/v1/ideas/{canvas_id}/edges", json=body)

    async def delete_edge(self, canvas_id: str, edge_id: str) -> dict[str, Any]:
        """Delete an edge from a canvas."""
        return await self._client._request("DELETE", f"/api/v1/ideas/{canvas_id}/edges/{edge_id}")

    # -------------------------------------------------------------------
    # Export & Promote
    # -------------------------------------------------------------------

    async def export_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Export a canvas as React Flow JSON."""
        return await self._client._request("GET", f"/api/v1/ideas/{canvas_id}/export")

    async def promote_nodes(self, canvas_id: str, node_ids: list[str]) -> dict[str, Any]:
        """Promote selected nodes from the idea canvas to the goals stage."""
        return await self._client._request(
            "POST",
            f"/api/v1/ideas/{canvas_id}/promote",
            json={"node_ids": node_ids},
        )
